anvil/observables.py

```diff
@@ -3,35 +3,63 @@
 
 functions for calculating observables on a stack of states.
 
-Functions
----------
-print_plot_observables:
-    can be considered the main function which in turn calculates the rest
-    of the observables defined in this module.}
-
 Notes
 -----
 Check the definitions of functions, most are defined according the the arxiv
 version: https://arxiv.org/pdf/1904.12072.pdf
 
 """
-from math import acosh
-
-import pandas as pd
-import matplotlib.pyplot as plt
 import numpy as np
 from scipy.signal import correlate
+import torch
+from tqdm import tqdm
 
-from reportengine.table import table
-from reportengine.figure import figure
+def arcosh(x):
+    """Inverse hyperbolic cosine function for torch.Tensor arguments.
 
+        arcosh(x) = log_e (x + sqrt(x^2 - 1) )
+                  = log_e (x) + log_e (1 + sqrt(x^2 - 1) / x)
+    """
+    c0 = torch.log(x)
+    c1 = torch.log1p(torch.sqrt(x * x - 1) / x)
+    return c0 + c1
 
-class GreenFunction:
-    def __init__(self, states, geometry):
+
+def bootstrap(observable):
+    """Return the standard deviation for an observable based on a number of
+    'bootstrap' samples."""
+    obs_full = observable[0]
```

Comment:
I don't really understand this convention, why is the 0th index something special, why not just have the boot samples [1:] as the full thing then take mean across samples and variance across samples?

Reply:
At some point I was planning to add a bias-correction step which would use the observable calculated on the 'full' sample as well as the mean of the bootstrap distribution. But the plots of bootstrap distributions of observables were all fairly symmetric so I didn't think this was urgent.

I guess for now I'll tidy it up and do as you say, and we can cross the bias bridge if we ever come to it!

Reply:
At some point I was planning to add a bias-correction step which would use the observable calculated on the 'full' sample as well as the mean of the bootstrap distribution. But the plots of bootstrap distributions of observables were all fairly symmetric so I didn't think this was urgent.

I guess for now I'll tidy it up and do as you say, and we can cross the bias bridge if we ever come to it!